tango/integrations/transformers/ia3.py

```diff
@@ -0,0 +1,244 @@
+    "facebook/opt-30b": OPT_IA3_CONFIG,
+    "facebook/opt-66b": OPT_IA3_CONFIG,
+    "EleutherAI/gpt-j-6B": GPT_J_IA3_CONFIG,
+}
+
+
+class LinearWithIA3(nn.Module):
+    def __init__(self, linear_layer, ia3_param_names, unfuse_size: int = None):
+        super().__init__()
+
+        self.in_features = linear_layer.in_features
+        self.out_features = linear_layer.out_features
+        self.unfuse_size = unfuse_size
+
+        self.weight = linear_layer.weight
+        self.bias = linear_layer.bias
+
+        self.ia3_param_names = ia3_param_names
+
+        # if (q,k,v) are stacked into one layer
+        if unfuse_size is not None:
+            assert linear_layer.out_features == unfuse_size * 3
+            # IA3 only operates on k and v (not q), thus the "* 2"
+            setattr(self, ia3_param_names, nn.Parameter(torch.ones(unfuse_size * 2, 1)))
+        else:
+            setattr(self, ia3_param_names, nn.Parameter(torch.ones(self.out_features, 1)))
+
+    def forward(self, x):
+        x = F.linear(x, self.weight, self.bias)
+
+        ia3_params = getattr(self, self.ia3_param_names)
+
+        if ia3_params.requires_grad:
+            if self.unfuse_size is not None:
+                # non_q means k and v
+                q, non_q = x[:, :, : self.unfuse_size], x[:, :, self.unfuse_size :]
+                ia3_params = getattr(self, self.ia3_param_names)
+                non_q = non_q * ia3_params.flatten()
+                x = torch.cat([q, non_q], dim=2)
+            else:
+                x = x * ia3_params.flatten()
+
+        return x
+
+
+class Conv1DWithIA3(nn.Module):
+    def __init__(self, conv1d_layer, ia3_param_names, unfuse_size: int = None):
+        super().__init__()
+
+        # nf: number of output features; nx: number of input features
+        self.nf = conv1d_layer.nf
+        self.unfuse_size = unfuse_size
+
+        self.weight = conv1d_layer.weight
+        self.bias = conv1d_layer.bias
+
+        self.ia3_param_names = ia3_param_names
+
+        # in c_att parameters, (q,k,v) linear layers are stacked into one Conv1D layer
+        if unfuse_size is not None:
+            assert conv1d_layer.nf == unfuse_size * 3
+            # but IA3 only operates on k and v (not q), thus the "* 2"
+            setattr(self, ia3_param_names, nn.Parameter(torch.ones(unfuse_size * 2, 1)))
+        else:
+            setattr(self, ia3_param_names, nn.Parameter(torch.ones(self.nf, 1)))
+
+    def forward(self, x):
+        # copied and pasted from the original Conv1D implemnetation
+        size_out = x.size()[:-1] + (self.nf,)
+        x = torch.addmm(self.bias, x.view(-1, x.size(-1)), self.weight)
+        x = x.view(size_out)  # ... * self.nf
+
+        ia3_params = getattr(self, self.ia3_param_names)
+
+        if ia3_params.requires_grad:
+            if self.unfuse_size is not None:
+                # non_q means k and v
+                q, non_q = x[:, :, : self.unfuse_size], x[:, :, self.unfuse_size :]
```

Comment:
Which transformer models have fused `Conv1D` layers? 

It's minor, but we can probably refactor away the code repetition.

Reply:
Of the architectures I made configs for here, only gpt2 uses `Conv1d` layers. Its attention layer is fused but its mlp is not so unfortunately we need both the `if self.unfuse-size is not None:` and `else:` control flow paths here.

Meanwhile bloom uses `nn.Linear` layers and also has a fused attention layer.

I could factor out the unfusing code here from the two classes. Or did you have something else in mind?

Reply:
Yes, that's what I had in mind.